tools/render-test/shader-input-layout.cpp

```diff
@@ -286,13 +286,54 @@ namespace renderer_test
                                 else if (parser.LookAhead("name"))
                                 {
                                     parser.ReadToken();
-                                    Token nameToken = parser.ReadToken();
 
+                                    // Optionally consume '=' 
+                                    if (parser.NextToken().Type == TokenType::OpAssign)
```

Comment:
Okay, at this point I decided to follow through and look at the `TokenReader` type that this stuff is using and... it is *not* something I'm happy with as a parser interface. I guess it is only used for parsing these test input layout things, so it isn't the end of the world, but just for the record, here's the operations I'd want a good low-level parser interface to support:

* `peekToken()` should give read-only access to the next token ahead of the cursor (which is an "end-of-file" token if there are no more), while `peekTokenCode()` should give access to the token code ("type" for this `TokenReader`)

* `readToken()`, `getToken()`, or `advanceToken()` should consume and advance one token (or stay put on end-of-file) without any errors

* `readIf(code)` or `advanceIf(code)` should basically do `if(peekCode()==code) readToken();` and return a `bool` to indicate if anything was read (bonus points if you make the `Token` type convertible to `bool` and have an "invalid"/`false` case that you return for operations like this, with all other tokens being "truthy"). Many of the operations in this parsing logic could be replaced with `readIf`/`advanceIf` operations.

* `expect(code)` should behave like `readIf`/`advanceIf` with the added step of emitting an error message if the next token doesn't match.

* Note that `expect(TokenType::EndOfFile)` should Just Work with the above definitions, which is actually quite helpful

Rather than immediately throw an exception and halt parsing (as this `TokenReader` does), operations like `expect` should set the parser into a "recovery" state (this can be a simple `bool` in practice). In the recovery state, any `readIf`/`expect` sort of operation that succeeds should exit the recovery state, while any `expect` that fails should *not* emit an error and should instead scan forward through the input looking for a candidate matching token (throwing out input along the way). The exact behavior of what can/should be skipped in recovery is the hardest part to tune, but the hope is to avoid cascading errors while also parsing as much of the input as possible in one pass.

I don't expect us to actually change `TokenReader` for now, but having it in the `core` libraries carries an implication that it is actually a high-quality interface for doing its chosen job and I just don't buy it.

(Note: none of this really related to this PR, which is just doing the best it can with a pre-established interface)

Reply:
Agreed it is not great, I was just using what was used elsewhere. It would be better if I could make requests to the Slang reflection, but I didn't see that supported. 

Also note another aspect of this class is that it has the same class name (and is in the namespace) as the main slang lexer which can confuse IDs.

Reply:
Yeah, the name overlap is really unfortunate. In the short term we could rename one or the other, but eventually we might want to re-introduce a bit more refined namespacing back into the project to help with stuff like this (where if makes sense to have a basic version of something in core, but the actual compiler needs a much more advanced version of the same kind of thing)